src/main/java/jenkins/branch/WorkspaceLocatorImpl.java

```diff
@@ -305,6 +305,16 @@ public Void invoke(File f, VirtualChannel channel) throws IOException, Interrupt
         }
     }
 
+    private static final Map<Node, Object> nodeLocks = new WeakHashMap<>();
```

Comment:
I have seen some things elsewhere that explicitly do not use HashMaps for Nodes because the node hash() is not stable.  (e,.g. changing a name or something changes the actual hash).
in other words needs to really be a `WeakIdentityHashMap` which doesn't exist in stock JDK

Reply:
NB the [`Slave`](https://github.com/jenkinsci/jenkins/blob/3f15ad20985066661158ee1bb0b3255e000b3081/core/src/main/java/hudson/model/Slave.java#L552-L565) subclass is specifically safe for this - and the `Node` class does not implement hashCode so delegates to `Object` which is safe. however not all implementations may be

Reply:
`Node` itself does not override `hashCode` or `equals`. `Slave` does, based on `.name` which is technically mutable but I think only ever actually set in the constructor or during deserialization: renaming a `Slave` would create a new instance I believe, via `@DataBoundConstructor`.

I would rather base this on object identity if something like `java.beans.WeakIdentityMap` were `public`, but I think it is OK since if you in fact reconfigure the agent (so that the current patch uses the same lock as before) there is no real harm done because the index cache is keyed off of `Channel` identity, which will have changed.

Reply:
> not all implementations may be

Theoretically. I could select a lock based on the node name I guess, just in case.